CMakeLists.txt

```diff
@@ -2,22 +2,48 @@
 cmake_minimum_required(VERSION 3.4)
 project(libuv LANGUAGES C)
 
+cmake_policy(SET CMP0057 NEW)
+cmake_policy(SET CMP0063 NEW)
+cmake_policy(SET CMP0064 NEW)
+
+list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake")
+
 include(CMakePackageConfigHelpers)
 include(CMakeDependentOption)
+include(CheckCCompilerFlag)
 include(GNUInstallDirs)
 include(CTest)
 
+set(CMAKE_C_VISIBILITY_PRESET hidden)
+set(CMAKE_C_STANDARD_REQUIRED ON)
+set(CMAKE_C_EXTENSIONS ON)
+set(CMAKE_C_STANDARD 90)
+
 cmake_dependent_option(LIBUV_BUILD_TESTS
   "Build the unit tests when BUILD_TESTING is enabled and we are the root project" ON
   "BUILD_TESTING;CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR" OFF)
 
+cmake_dependent_option(LIBUV_BUILD_BENCH
+  "Build the benchmarks when building unit tests and we are the root project" ON
+  "LIBUV_BUILD_TESTS" OFF)
+
+check_c_compiler_flag(-Wstrict-prototypes UV_LINT_STRICT_PROTOTYPES)
+check_c_compiler_flag(-Wextra UV_LINT_EXTRA)
+check_c_compiler_flag(-Wall UV_LINT_DEFAULT)
+
+check_c_compiler_flag(-Wno-unused-parameter UV_LINT_NO_UNUSED_PARAMETER)
+
+set(lint-strict-prototypes $<$<BOOL:${UV_LINT_STRICT_PROTOTYPES}>:-Wstrict-prototypes>)
```

Comment:
Where can I read some more on this syntax?

Reply:
These are CMake's [generator expressions][]. The best analogy I can give is that they are a very poor imitation of s-expressions. One thing not mentioned is that they do not support being written on multiple lines in place or in a call to set. The "best" way to write a multi-line generator expression is via `string(CONCAT)`. I have a few examples in an upcoming PR when we need to use an `$<OR>` expression, but I want to keep it readable.

Generator expressions are most useful in that they allow you to remove a majority of your configure time branching and push them into generation time. Configuration in CMake is single-threaded, while generation is multi-threaded.

[generator expressions]: https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html

Reply:
> they do not support being written on multiple lines in place or in a call to set

Does this work?
```cmake
set(lint-no-unused-parameter
     $<$<BOOL:${UV_LINT_NO_UNUSED_PARAMETER}>:-Wno-unused-parameter)
```
I'd prefer to keep lines <= 80 columns if at all possible.